File: src/simulation/runner.js
// Play one or more complete games between two bots and return results.
import { createBoard, place } from '../game/board.js';
import { checkWin, isFull } from '../game/rules.js';
import { AI, HUMAN } from '../game/constants.js';

/**
 * @typedef {Object} GameResult
 * @property {'A'|'B'|null} winner  - which bot won, or null for draw
 * @property {number}       moves   - total plies played
 * @property {number}       duration - wall-clock ms
 */

/**
 * Run a single game between two bot functions.
 *
 * Token assignment: the first player always uses AI (2), second uses HUMAN (1).
 * `botAGoesFirst` controls which bot is the first player.
 *
 * @param {Function} botA  - (board, token) => colIndex
 * @param {Function} botB  - (board, token) => colIndex
 * @param {object}   opts
 * @param {boolean}  opts.botAGoesFirst  - default true
 * @returns {GameResult}
 */
export function runGame(botA, botB, { botAGoesFirst = true } = {}) {
  const board = createBoard();

  // bots[0] is the first player (token AI), bots[1] is the second (token HUMAN)
  const bots   = botAGoesFirst ? [botA, botB] : [botB, botA];
  const tokens = [AI, HUMAN];

  let moves = 0;
  const start = Date.now();

  while (true) {
    const turn  = moves % 2;
    const token = tokens[turn];
    const col   = bots[turn](board, token);

    place(board, col, token);
    moves++;

    if (checkWin(board, token)) {
      // turn 0 is first player; botA is first player iff botAGoesFirst
      const winnerIsA = (turn === 0) === botAGoesFirst;
      return { winner: winnerIsA ? 'A' : 'B', moves, duration: Date.now() - start };
    }

    if (isFull(board)) {
      return { winner: null, moves, duration: Date.now() - start };
    }
  }
}

/**
 * Run N games between botA and botB, alternating who goes first each game.
 * Returns win/loss/draw counts from botA's perspective.
 *
 * @param {Function} botA
 * @param {Function} botB
 * @param {number}   n
 * @returns {{ wins: number, losses: number, draws: number }}
 */
export function runGames(botA, botB, n = 100) {
  let wins = 0, losses = 0, draws = 0;

  for (let i = 0; i < n; i++) {
    const { winner } = runGame(botA, botB, { botAGoesFirst: i % 2 === 0 });
    if (winner === 'A')   wins++;
    else if (winner === 'B') losses++;
    else                     draws++;
  }

  return { wins, losses, draws };
}

File: src/simulation/runner.py
# Play one or more complete games between two bots and return results.
import time
from ..game.board import create_board, place
from ..game.rules import check_win, is_full
from ..game.constants import AI, HUMAN


def run_game(bot_a, bot_b, bot_a_goes_first=True):
    """
    Run a single game between two bot functions.

    Token assignment: the first player always uses AI (2), second uses HUMAN (1).
    `bot_a_goes_first` controls which bot is the first player.

    Returns dict: { 'winner': 'A' | 'B' | None, 'moves': int, 'duration': float (ms) }
    """
    board = create_board()

    bots   = [bot_a, bot_b] if bot_a_goes_first else [bot_b, bot_a]
    tokens = [AI, HUMAN]

    moves = 0
    start = time.perf_counter()

    while True:
        turn  = moves % 2
        token = tokens[turn]
        col   = bots[turn](board, token)

        place(board, col, token)
        moves += 1

        if check_win(board, token):
            winner_is_a = (turn == 0) == bot_a_goes_first
            duration = (time.perf_counter() - start) * 1000
            return {'winner': 'A' if winner_is_a else 'B', 'moves': moves, 'duration': duration}

        if is_full(board):
            duration = (time.perf_counter() - start) * 1000
            return {'winner': None, 'moves': moves, 'duration': duration}


def run_games(bot_a, bot_b, n=100):
    """
    Run N games between bot_a and bot_b, alternating who goes first each game.
    Returns win/loss/draw counts from bot_a's perspective.

    Returns dict: { 'wins': int, 'losses': int, 'draws': int }
    """
    wins = losses = draws = 0

    for i in range(n):
        result = run_game(bot_a, bot_b, bot_a_goes_first=(i % 2 == 0))
        if result['winner'] == 'A':
            wins += 1
        elif result['winner'] == 'B':
            losses += 1
        else:
            draws += 1

    return {'wins': wins, 'losses': losses, 'draws': draws}
